docs/design/4940-reliable-loki-pipelines.md

```diff
@@ -0,0 +1,153 @@
+
+We could adopt the same pattern for loki pipelines as well with the following interface:
+
+```go
+type Consumer interface {
+	Consume(ctx context.Context, entries []Entry) error
+}
+```
+
+Adopting this pattern for loki pipelines would change it from a channel-based pipeline to a function-based pipeline. This would give us two things:
+1. Increased throughput because several sources such as many files or http requests can now call the next component in the pipeline at the same time.
+2. A way to return signals back to the source so we can handle things like giving a proper error response or determine if the position file should be updated. 
+
+Solving the issues listed above.
+
+A batch of entries should be considered successfully consumed when they are queued up for sending. We could try to extend this to when it was successfully sent over the wire, but that could be considered an improvement at a later stage.
+
+Pros:
+* Increase throughput of log pipelines.
+* A way to signal back to the source
+Cons:
+* We need to rewrite all loki components with this new pattern and make them safe to call in parallel.
+* We go from an iterator-like pipeline to passing slices around. Every component would have to iterate over this slice and we need to make sure it's safe to mutate because of fan-out.
+
+## Proposal 2: Appendable
+
+The prometheus pipeline uses [Appendable](https://github.com/prometheus/prometheus/blob/main/storage/interface.go#L62).
+Appendable only has one method `Appender` that will return an implementation of [Appender](https://github.com/prometheus/prometheus/blob/main/storage/interface.go#L270).
+
+We could adopt this pattern for loki pipelines by having:
+```go
+type Appendable interface {
+    Appender(ctx context.Context) Appender
+}
+
+type Appender interface {
+    Append(entry Entry) error
+    Commit() error
+    Rollback() error
+}
+```
+
+This approach would, like Proposal 1, solve the issues listed above with a function-based pipeline, but the pipeline would still be iterator-like (one entry at a time).
+
+### How it works
+Source components would: 
+Obtain an `Appender` that can fan-out to all downstream components, then call `Append` for each entry.
+If every call to `Append` is successful, `Commit` should be called; otherwise `Rollback`.
+
+Processing components would:
+Implement `Appendable` to return an `Appender` that runs processing for each entry and fan-out to all downstream components.
+
+Sink components would:
+Implement `Appendable` to return an `Appender` that buffers entries until either `Commit` or `Rollback` is called.
+
+Pros:
+* Increase throughput of log pipelines.
+* A way to signal back to the source
+* Iterator-like pipeline - one entry at a time
+* Transaction semantics provide better error handling
+* Aligns with Prometheus patterns already in use
```

Comment:
I'm also not sure this is a pro 😅 Prometheus patterns come from a database world, we are a telemetry collector.